compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -11946,27 +11951,36 @@ private boolean isNestedQueryExpr() {
         return Collections.frequency(this.errorHandler.getContextStack(), ParserRuleContext.QUERY_EXPRESSION) > 1;
     }
 
-    private boolean isValidIntermediateQueryStart(SyntaxKind syntaxKind) {
-        switch (syntaxKind) {
+    private boolean isValidIntermediateQueryStart(STToken token) {
+        switch (token.kind) {
             case FROM_KEYWORD:
             case WHERE_KEYWORD:
             case LET_KEYWORD:
             case SELECT_KEYWORD:
-            case COLLECT_KEYWORD:
             case JOIN_KEYWORD:
             case OUTER_KEYWORD:
             case ORDER_KEYWORD:
-            case GROUP_KEYWORD:
             case BY_KEYWORD:
             case ASCENDING_KEYWORD:
             case DESCENDING_KEYWORD:
             case LIMIT_KEYWORD:
                 return true;
+            case IDENTIFIER_TOKEN:
+                return isGroupOrCollectKeyword(token);
             default:
                 return false;
         }
     }
 
+    private static boolean isGroupOrCollectKeyword(STToken nextToken) {
+        if (nextToken.kind != SyntaxKind.IDENTIFIER_TOKEN || nextToken.isMissing()) {
+            return false;
+        }
+        String tokenText = ((STIdentifierToken) nextToken).text;
+        return tokenText.equals(SyntaxKind.COLLECT_KEYWORD.stringValue())
+                || tokenText.equals(SyntaxKind.GROUP_KEYWORD.stringValue());
```

Comment:
Seems there is an easy way.
Let's add a method as follows to check if an identifier is a given SyntaxKind keyword.

```java
    private static boolean isKeywordMatch(SyntaxKind syntaxKind, STToken token) {
        return token.kind == SyntaxKind.IDENTIFIER_TOKEN && syntaxKind.stringValue().equals(token.text());
    }
```